app/domain/authentication/authn_gcp/update_authenticator_input.rb

```diff
@@ -0,0 +1,105 @@
+require 'command_class'
+
+module Authentication
+  module AuthnGcp
+
+    UpdateAuthenticatorInput = CommandClass.new(
+      dependencies: {
+        verify_and_decode_token: Authentication::OAuth::VerifyAndDecodeToken.new,
+        decoded_token_class:     DecodedToken,
+        logger:                  Rails.logger
+      },
+      inputs:       [:authenticator_input]
+    ) do
+
+      extend Forwardable
+      def_delegators :@authenticator_input, :authenticator_name, :account, :credentials, :username
+
+      def call
+        validate_token
+        validate_conjur_username
+        updated_input
+      end
+
+      private
+
+      def validate_token
+        decoded_token
+        validate_audience
+      end
+
+      def decoded_token
+        @decoded_token ||= @decoded_token_class.new(
+          decoded_token_hash: @verify_and_decode_token.call(
+            provider_uri:     PROVIDER_URI,
+            token_jwt:        decoded_credentials.jwt,
+            claims_to_verify: {
+              verify_iss:        true,
+              iss:               ISS,
+              verify_iat:        true,
+              verify_expiration: true
+            }
+          ),
+          logger:             @logger
+        )
+      end
+
+      def decoded_credentials
+        @decoded_credentials ||= Authentication::Jwt::DecodedCredentials.new(credentials)
+      end
+
+      def validate_audience
+        if audience_parts.length != 3 ||
+          audience_parts[0] != "conjur" ||
+          audience_parts[1] != account
+          raise Errors::Authentication::AuthnGcp::InvalidAudience, audience
+        end
+      end
+
+      def audience_parts
+        @audience_parts ||= audience.split('/')
+      end
+
+      def audience
+        @audience ||= @decoded_token.audience
+      end
+
+      def validate_conjur_username
+        if conjur_username == "admin"
+          raise Errors::Authentication::AdminAuthenticationDenied, authenticator_name
```

Comment:
This is an interesting point, do you think we should block admin? I think we should take it with @sashaCher

Reply:
does it make sense for the admin to authenticate with `authn-gcp`? seems like a backdoor to me.

Reply:
Thank you for raising it.

I think that `admin` should not have an ability to authenticate using `authn-gcp` unless it is explicitly defined. Means somebody has added `admin` to the `consumers` group.

But as far as I understand an authenticated role has to have at least one `gcp` related annotations, hasn't it? Why does`admin` demand so special treatment?

Reply:
i thought about it more and it doesn't need a special care. i took this from `authn-oidc` but there we had another issue where we didn't want the OIDC Provider admin to be authenticated as the Conjur admin. Here it's not really an issue and because we need at least one gcp annotation we can let the admin pass at this stage.


thanks!